### DISCUSSION TITLE
(0, someFunction)() in javascript

### DISCUSSION THREAD

**Discussion by mediumhust** [TinyHust/notes#262] [Vault]
```javascript
(function() {
  (0,eval)("var foo = 123"); // indirect call to eval, creates global variable
})();
console.log(foo);            // 123
(function() {
  eval("var bar = 123");     // direct call to eval, creates local variable
})();
console.log(bar);            // ReferenceError
```
REF: https://2ality.com/2015/12/references.html

**Comment by proofmancer** [comment]
The code you provided demonstrates the difference between using indirect and direct calls to eval() in JavaScript. The key difference is whether the eval() call is performed directly or indirectly. Let's break down the code and its behavior:

```
(function() {
  (0,eval)("var foo = 123"); // indirect call to eval, creates global variable
})();
console.log(foo);            // 123

```

In this code, an indirect call to eval() is made using (0, eval)("..."). This approach is used to prevent some of the scoping behavior associated with direct eval() calls. Since the call is indirect, it creates a global variable foo because the eval() function is executed in the global scope.

```
(function() {
  eval("var bar = 123");     // direct call to eval, creates local variable
})();
console.log(bar);            // ReferenceError

```

In this code, a direct call to eval() is made. The eval() function is executed in the local scope of the immediately-invoked function expression (IIFE), creating a local variable bar. Since bar is local to the IIFE's scope, attempting to access it outside the IIFE results in a ReferenceError because it's not defined in the outer scope.

This behavior highlights the importance of avoiding the use of eval() whenever possible, as it can lead to unexpected scoping and security issues. The code also demonstrates how indirect calls to eval() can behave differently from direct calls in terms of variable scope.
